### PR TITLE
F#232 purging rt unflriendly operations

### LINKED ISSUE #232: Purge real time unfriendly operations from SPI controllers
Some of the SPI related classes contain realtime unfriendly operations that are called from the `update()` functions of the controllers.

The ones detected so far are related to possible memory allocation. Memory allocation should be banned form real time parts of the code to make it more deterministic.
- push and pop operations on several queues could result in allocation of new chunks of memory to the queue.

A suggestion to fix this is to use boost::circular_buffer with a fixed size as a container for the queue (instead of the default deque) as suggested [here](http://stackoverflow.com/questions/1303993/pre-allocate-space-for-c-stl-queue). Of course the size of the queue should be enough that we trust it won't be exceeded.
- use of the  `new` operator is forbidden in the realtime thread. It is being used [here](https://github.com/shadow-robot/sr-ronex/blob/indigo-devel/sr_ronex_controllers/src/spi_sensor_read_controller.cpp#L127) and it also causes a **memory leak**.

### LINKED ISSUE #233: Fix ronex passthrough controller
In the service callback [here](https://github.com/shadow-robot/sr-ronex/blob/indigo-devel/sr_ronex_controllers/src/spi_passthrough_controller.cpp#L87) we shouldn't be looping through all the SPI port's queues, but only accessing the one that corresponds to `spi_out_index`.

### LINKED ISSUE #234: Fix spi_base_controller update function
In spi_base_controller, the variable `new_command` should be a vector. Currently it is used inside a loop through the spi channels [here](https://github.com/shadow-robot/sr-ronex/blob/indigo-devel/sr_ronex_controllers/src/spi_base_controller.cpp#L103), and it affects all the channels without distinction.

### PR DESCRIPTION
**Pull request by beatrizleon**
Fixes #232
Fixes #233 
Fixes #234

### CODE REVIEW

File: sr_ronex_controllers/include/sr_ronex_controllers/spi_base_controller.hpp [hunk 1]
```diff
@@ -76,8 +86,8 @@ class SPIBaseController
 
   ronex::SPI* spi_;
 
-  std::vector<std::queue<SplittedSPICommand*> > command_queue_;
-  std::vector<std::queue<std::pair<SplittedSPICommand*, SPI_PACKET_IN* > > > status_queue_;
+  std::vector<std::queue<int, boost::circular_buffer<SplittedSPICommand> > > command_queue_;
```
Comment (toliver) [inline sr_ronex_controllers/include/sr_ronex_controllers/spi_base_controller.hpp]: Shouldn't this be:

``` cpp
std::vector<std::queue<SplittedSPICommand, boost::circular_buffer<SplittedSPICommand> > > command_queue_;
```

Comment (beatrizleon) [inline sr_ronex_controllers/include/sr_ronex_controllers/spi_base_controller.hpp]: Fixed

File: sr_ronex_controllers/src/spi_sensor_read_controller.cpp [hunk 2]
```diff
@@ -109,39 +109,36 @@ void SPISensorReadController::update(const ros::Time& time, const ros::Duration&
...
-        if (new_command)
-        {
-          new_command = false;
-          spi_->nullify_command(*channel_iter);
-        }
-
         // the response has not been received. If the command type is NORMAL
         // then the response can be updated (it's INVALID until the SPI responds)
-        if (spi_->state_->command_type == RONEX_COMMAND_02000002_COMMAND_TYPE_NORMAL)
+        if (spi_->state_->command_type == RONEX_COMMAND_02000002_COMMAND_TYPE_NORMAL )
```
Comment (toliver) [inline sr_ronex_controllers/src/spi_sensor_read_controller.cpp]: Extra space, why?

File: sr_ronex_controllers/src/spi_passthrough_controller.cpp [hunk 3]
```diff
@@ -84,40 +84,37 @@ bool SPIPassthroughController::command_srv_cb(sr_ronex_msgs::SPI::Request &req,
...
-
-            // we used the status (sent it back to the user through the service
-            // response -> popping from the queue
-            status_queue_[i].pop();
-
-            break;
-          }
+          hex << static_cast<unsigned int>(status_queue_[spi_out_index].front().second.packet.data_bytes[j]);
         }
+          catch(...)
```
Comment (toliver) [inline sr_ronex_controllers/src/spi_passthrough_controller.cpp]: Bad indentation for catch?

File: sr_ronex_controllers/src/spi_base_controller.cpp [hunk 4]
```diff
@@ -95,50 +106,55 @@ void SPIBaseController::update(const ros::Time&, const ros::Duration&)
...
-
         // the response has not been received. If the command type is NORMAL
         // then the response can be updated (it's INVALID until the SPI responds)
         if ( spi_->state_->command_type == RONEX_COMMAND_02000002_COMMAND_TYPE_NORMAL )
         {
-          status_queue_[spi_index].front().second =
-                  new SPI_PACKET_IN(spi_->state_->info_type.status_data.spi_in[spi_index]);
+          status_queue_[spi_index].back().second.received = true;
+          status_queue_[spi_index].back().second.packet =
+              SPI_PACKET_IN(spi_->state_->info_type.status_data.spi_in[spi_index]);
```
Comment (toliver) [inline sr_ronex_controllers/src/spi_base_controller.cpp]: I think we don't need to use the constructor here, as the assignment will do a copy of the whole structure anyway. We don't need an intermediate copy.

``` cpp
status_queue_[spi_index].back().second.packet = spi_->state_->info_type.status_data.spi_in[spi_index];
```

**Pull request merged**